### PR TITLE
Update C++ standard to 14 and clean up

### PR DESCRIPTION
**Pull request by adamemerson**
Still an improvement!

### CODE REVIEW

File: src/common/backport_std.h [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+ *
+ */
+
+#include <memory>
+#include <type_traits>
+
+#ifndef CEPH_COMMON_BACKPORT14_H
+#define CEPH_COMMON_BACKPORT14_H
+
+// Library code from C++14 that can be implemented in C++11.
```
Comment (cbodley) [inline src/common/backport_std.h]: nit: this comment can be updated

File: src/common/backport14.h [hunk 2]
```diff
@@ -21,18 +21,119 @@
 // Library code from C++14 that can be implemented in C++11.
 
 namespace ceph {
-template<typename T>
-using remove_extent_t = typename std::remove_extent<T>::type;
-template<typename T>
-using remove_reference_t = typename std::remove_reference<T>::type;
-template<typename T>
-using result_of_t = typename std::result_of<T>::type;
+// Template variable aliases from C++17 <type_traits>
```
Comment (tchaikov) [inline src/common/backport14.h]: can we guard this with `#if __cplusplus < 201703L`, and just put these templates into `namespace std`?

**Comment by adamemerson** [comment]
@tchaikov Maaaaybe? It's Undefined Behavior and Bad and Naughty and so I'd rather not, but it would probably work.

I have a /better/ solution for this sort of problem I've been working on from time to time.

Generally, if we put everything in core inside the ceph namespace

And encapsulated things like this in a few inline namespaces so we could easily pull them into RGW and RBD and whatnot

Then we could easily change definitions to usings and things like that, but it's a fairly large project and I haven't had a chance to work on it recently.

**Comment by cbodley** [comment]
> It's Undefined Behavior and Bad and Naughty and so I'd rather not

that's my feeling as well. it might make our transitions between c++ versions a bit easier, but i don't think we do that often enough to get so adventurous

**Comment by adamemerson** [comment]
Yeah, once we go to 17, whenever the CentOS people get their thing done, we won't have to change this sort of thing until …2020?

**Comment by adamemerson** [comment]
@tchaikov One minor error crept in during a rebase, just fixed it.

**Comment by adamemerson** [comment]
Well, major in that it broke everything, but minor in terms of scope and fix and complexity.

File: src/common/backport14.h [hunk 3]
```diff
@@ -1,212 +0,0 @@
...
-
-// Library code from C++14 that can be implemented in C++11.
-
-namespace ceph {
-template<typename T>
-using remove_extent_t = typename std::remove_extent<T>::type;
-template<typename T>
-using remove_reference_t = typename std::remove_reference<T>::type;
-template<typename T>
-using result_of_t = typename std::result_of<T>::type;
```
Comment (tchaikov) [inline src/common/backport14.h]:
```
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/dout.h:22:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/common/debug.h:18,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/compressor/Compressor.cc:22:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/config.h:171:11: error: 'result_of_t' in namespace 'ceph' does not name a template type
     ceph::result_of_t<Callback(const T&, Args...)> {
           ^~~~~~~~~~~
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/config.h:171:22: error: expected initializer before '<' token
     ceph::result_of_t<Callback(const T&, Args...)> {
                      ^
src/compressor/CMakeFiles/compressor_objs.dir/build.make:62: recipe for target 'src/compressor/CMakeFiles/compressor_objs.dir/Compressor.cc.o' failed
```

@adamemerson

File: src/test/common/test_backport14.cc [hunk 4]
```diff
@@ -14,7 +14,7 @@
  *
  */
 
-#include "common/backport14.h" // include first: tests that header is standalone
```
Comment (tchaikov) [inline src/test/common/test_backport14.cc]: @adamemerson we need to rename this source file. see https://github.com/ceph/ceph/pull/19490/files#diff-a67cc84c25b8865adfc63fed42d6ceffR268

**Comment by adamemerson** [comment]
Oops. My own merged patch did me in. Fixed and pushed. Will run a local compile to check anything else.

**Comment by adamemerson** [comment]
@tchaikov All right, got the problems I introduced by rebasing cleared away.

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2017-12-15_11:03:39-rados-wip-kefu-testing-2017-12-15-1528-distro-basic-mira/
- http://pulpito.ceph.com/kchai-2017-12-16_03:28:12-rados-wip-kefu-testing-2017-12-15-1528-distro-basic-mira/

**Comment by cbodley** [comment]
exciting! 🎉 you are my heroes, @tchaikov and @adamemerson!

**Comment by chardan** [comment]
Very cool!!

**Comment by theanalyst** [comment]
Awesome!

**Comment by smithfarm** [comment]
@adamemerson @chardan Could you help me diagnose the build failure we are seeing in nautilus? Here are two instances of it:

https://github.com/ceph/ceph/pull/30049
https://github.com/ceph/ceph/pull/30050

It blurts out a gazillion errors, but the first one is:

```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/convenience.h:46:24: error: 'result_of_t' is not a member of 'std'
   boost::optional<std::result_of_t<F(const std::decay_t<T>)>>
```

which would appear to be related to this PR?

Also, at https://en.cppreference.com/w/cpp/types/result_of I saw that this feature added by C++14 is now deprecated and due to be removed in C++20... not sure what should be done (if anything) about that.

**Comment by smithfarm** [comment]
@tchaikov Looks like "make check" is broken in nautilus? (See the previous commit - all of the "make check" failures in nautilus seem to be caused by this same issue, i.e. ostensibly "'result_of_t' is not a member of 'std'"...)

**Comment by tchaikov** [comment]
@smithfarm thanks for looking into this. yeah, i am afraid "rëtest this please" won't work this time. will investigate this early tomorrow.

**Comment by chardan** [comment]
@smithfarm Yeah, for C++17 result_of was deprecated, and removed in C++20 (see the "resolved issues" section here: https://tinyurl.com/km5tkhu). Chances are good that you can just replace it with "std::invoke_result" and "std::invoke_result_t".

**Comment by chardan** [comment]
I also saw the earlier part of the conversation related to feature testing, and wanted to point out that there are lots of new feature-test macros made available:
https://en.cppreference.com/w/cpp/feature_test
...and a good article from Study Group 10 on using them:
https://isocpp.org/std/standing-documents/sd-6-sg10-feature-test-recommendations

**Comment by tchaikov** [comment]
@smithfarm i failed to reproduce the FTBFS in my own xenial container and in #30083 . the only thing i can think of is to use GCC-8. see #30089.

but i think the root cause is still unknown. as the error message put
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/common/ceph_time.h:504:16: warning: variable templates only available with -std=c++14 or -std=gnu++14
 constexpr bool converts_to_timespec_v = converts_to_timespec<Clock>::value;
                ^~~~~~~~~~~~~~~~~~~~~~
```
and we do pass `-std=c++17` to g++. see https://github.com/ceph/ceph/blob/e1f67109056f9a6137e018badb8c923833e8eeba/src/CMakeLists.txt#L115

**Comment by cbodley** [comment]
in the VERBOSE output, i see that an extra `-std=gnu++11` is getting sprinkled into some targets after the `-std=c++17`. is it possible that cmake is getting confused by the `set(CMAKE_CXX_STANDARD 11)`s in src/dmclock/CMakeLists.txt and src/fmt/support/cmake/cxx14.cmake? on master, we have `set(CMAKE_CXX_STANDARD 17)`, but here we're just adding `-std=c++17` to `CMAKE_CXX_FLAGS`

**Comment by tchaikov** [comment]
@cbodley thanks a lot! that explains. i ran into the same issue couple weeks ago before switching over to cmake 3.10.

i believe that's the root cause.

**Comment by tchaikov** [comment]
but we need to either use cmake 3.10 for addressing this issue or hack libfmt to prevent it from falling back to C++11.

**Comment by cbodley** [comment]
what do we need from cmake 3.10? is that what adds 17 as an option for CXX_STANDARD?

**Comment by tchaikov** [comment]
yeah, 3.8 introduced the support of CXX_STANDARD=17.

**Comment by tchaikov** [comment]
i am working on a fix for libfmt and dmclock.

**Comment by tchaikov** [comment]
#30113 #30114

**Pull request merged**